lib/Pulse.Lib.AVLTree.fst

```diff
@@ -0,0 +1,948 @@
+{
+  fold (is_tree x T.Leaf); 
+}
+```
+
+```pulse
+ghost
+fn elim_is_tree_node (#t:Type0) (ct:tree_t t) (data:t) (ltree:T.tree t) (rtree:T.tree t)
+requires is_tree ct (T.Node data ltree rtree)
+ensures (
+  exists* (p:node_ptr t) (lct:tree_t t) (rct:tree_t t).
+    pure (ct == Some p) **
+    pts_to p { data; left = lct;right = rct } **
+    is_tree lct ltree **
+    is_tree rct rtree
+)
+{
+  unfold is_tree
+}
+```
+
+module G = FStar.Ghost
+
+```pulse
+ghost
+fn intro_is_tree_node (#t:Type0) (ct:tree_t t) (v:node_ptr t) (#node:node t) (#ltree:T.tree t) (#rtree:T.tree t)
+requires
+  pts_to v node **
+  is_tree node.left ltree **
+  is_tree node.right rtree **
+  pure (ct == Some v)
+ensures
+  is_tree ct (T.Node node.data ltree rtree)
+{
+  fold (is_tree ct (T.Node node.data ltree rtree))
+}
+```
+
+//is_tree_cases$
+let is_tree_cases #t x ft
+= match x with
+  | None -> pure (ft == T.Leaf)
+  | Some v -> 
+    exists* (n:node t) (ltree:T.tree t) (rtree:T.tree t).
+      pts_to v n **
+      pure (ft == T.Node n.data ltree rtree) **
+      is_tree n.left ltree **
+      is_tree n.right rtree
+//is_tree_cases$
+
+```pulse //cases_of_is_tree$ -CHECK THIS AGAIN
+ghost
+fn cases_of_is_tree #t (x:tree_t t) (ft:T.tree t)
+requires is_tree x ft
+ensures  is_tree_cases x ft
+{
+  match ft {
+    T.Leaf -> {
+      unfold (is_tree x T.Leaf);
+      fold (is_tree_cases None ft);
+    }
+    T.Node data ltree rtree -> {
+      unfold (is_tree x (T.Node data ltree rtree));
+      with p _lct _rct. _;
+      fold (is_tree_cases (Some p) ft)
+    }
+  }
+}
+
+```
+
+```pulse //is_list_case_none$
+ghost
+fn is_tree_case_none (#t:Type) (x:tree_t t) (#l:T.tree t)
+requires is_tree x l ** pure (x == None)
+ensures  is_tree x l ** pure (l == T.Leaf)
+{
+  cases_of_is_tree None l;
+  unfold is_tree_cases;
+  intro_is_tree_leaf x;
+}
+
+```
+
+```pulse //is_list_case_some$
+ghost
+fn is_tree_case_some (#t:Type) (x:tree_t t) (v:node_ptr t) (#ft:T.tree t) 
+requires is_tree x ft ** pure (x == Some v)
+ensures  
+   exists* (node:node t) (ltree:T.tree t) (rtree:T.tree t).
+    pts_to v node **
+    is_tree node.left ltree **
+    is_tree node.right rtree **
+    pure (ft == T.Node node.data ltree rtree)
+  
+{
+  cases_of_is_tree (Some v) ft;
+  unfold is_tree_cases;
+}
+```
+
+///////////////////////////////////////////////////////////////////////////////
+
+```pulse //height$
+fn rec height (#t:Type0) (x:tree_t t)
+requires is_tree x 'l
+returns n:nat
+ensures is_tree x 'l ** pure (n == T.height 'l)
+{
+   match x {
+    None -> {
+       is_tree_case_none x;
+       0
+    }
+    Some vl -> {
+      is_tree_case_some x vl;
+      let node = !vl;
+      let l_height = height node.left;
+      let r_height = height node.right;
+      intro_is_tree_node x vl;
+      if (l_height > r_height) {
+        (l_height + 1)
+      } else {
+        (r_height + 1)
+      }
+    }
+  }
+}
+```
+
+```pulse
+fn add_k_height (#t:Type0) (x:tree_t t) (k:nat)
```

Comment:
What is this add_k_height? Can we remove it?

Reply:
comment addressed.